### DISCUSSION TITLE
Enabling eager loading with custom `WHERE`/`LIMIT`/etc...

### DISCUSSION THREAD

**Discussion by IohannesArnold** [sqlalchemy/sqlalchemy#9898] [Ideas]
(Still a relative beginner at SQLAlchemy, so entirely possible I'm still missing something.)

I'm trying to eager load as much as possible, but sometimes I want to filter, set a LIMIT to, or change the ORDER BY for the children in a parent's relationship, and I'm running into trouble.

[This issue](https://github.com/sqlalchemy/sqlalchemy/issues/5053) also proposes adding functionality to allow filtered eager loads, but it issue is closed on the basis that "eagerly load partial collections is not a main use case in the SQLAlchemy ORM and it is generally not considered to be that great an idea, because that object you are creating is now in the Session and is not accurately representing the database state it claims to represent." This is understandable, and we can see someone getting confused by this point [on this mailing list thread](https://groups.google.com/g/sqlalchemy/c/lMji9T2mOgk/m/5BP_Q8B4DwAJ).

I do think there is one place where filtering an eager load would be helpful though, and that's for relationships built on `WriteOnlyCollection`. Since the attribute is currently never loaded at all, it's open for a filtered collection to get attached to, and since `WriteOnlyCollection` never claims to represent the state of the DB, it should not cause confusion for having a filtered collection.

I'm not familiar with the internals of SQLAlchemy and don't know what the most convenient API would look like. Perhaps this could be another method on `WriteOnlyCollection`; e.g. `WriteOnlyCollection.selectinload(limit=None,offest=None,order_by=None)`. Perhaps this could be a separate eager loading strategy; e.g. `Load.writeonlyload(attr: str | QueryableAttribute[Any])`. Maybe there's a better way; I don't know.

The reason I want to be able to eager load the relationship in this way is because I am writing a GraphQL API and there is enough information in a GraphQL query to eager load everything you need. So far I've been having a great time combining SQLAlchemy and a GraphQL library to eager load everything. But now I'm trying to implement filtering and limiting results for my API, and I can't figure out a way to do via eager loading. For example, to process the query below I can't get my backend to emit fewer than `COUNT(parents) +1` SELECT statements. In principle, with `selectin` loading, it should be exactly 3.

```graphql
{
  parents {
    children(first: 100) {
      grandchildren(first: 100) {
        field
      }
    }
  }
}

**Comment by zzzeek** [comment]
re: #5053, after many years of people asking for that feature, we eventually added it.   see [Adding Criteria to Loader Options](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#adding-criteria-to-loader-options).

however it wont change ORDER BY or add LIMIT.   you would still need to use explicit querying forms for things like that.

as for `WriteOnlyCollection`, there are no "loaders" with this collection type.  you only write your own queries for it and get results directly, just like querying for any other kind of rows.   so you can filter /limit /order by without restriction since "eager loaders" don't apply.

I think for your own case of loading hierarchies with open-ended filtering/limiting/ordering as well as arbitrary depth, you should forego trying to make "eager loaders" do this and just write a system that does the queries you want.    The style of eager loading used by subqueryload and selectinload was originally called "disjoint", and it's based on this recipe here: https://github.com/sqlalchemy/sqlalchemy/wiki/DisjointEagerLoading  - that approach can still work today.    Query for whatever objects you want and you can wire the lists of objects into a hierarchy using [attributes.set_committed_value()](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.attributes.set_committed_value).

**Reply by IohannesArnold** [reply]
Yes, I started using the strategy in "Adding Criteria to Loader Options" to filter relationships, but then I realized I wanted to set custom `LIMIT`s or `ORDER BY`, which, as you say, are not possible. I guess I'm looking for a loading option that's more powerful.

One thing that just now leads me to think of is `Load.with_expression()`, which is as general a loading option as possible. I suppose I could have a `query_expression` attribute for every relationship I wanted to load in this manner. Actually, would it be possible to use `Load.with_expression()` to populate a `WriteOnlyCollection`? If you could, then you wouldn't have to pollute a mapped class with a `query_expression` construct to accompany every relationship.

**Reply by zzzeek** [reply]
`WriteOnlyCollections` are not populated.  it's important to understand this.

**Reply by IohannesArnold** [reply]
Yes, I understand that at present `WriteOnlyCollections` are never populated. What I'm saying is, perhaps they ought to be (though granted that would make the name a misnomer)?

Or, if that's an obviously stupid suggestion, my more general point is that there seems to be a design space here which is not quite covered by existing apis. If someone says, "While loading several `parents`, I want to eagerly load their `children`, but subject to arbitrary `WHERE`/`LIMIT`/`ORDER BY`/etc. How do I do that?", then it doesn't seem there is a clear right tool. `PropComparator.and_()` will filter but won't do limits, etc. The principle behind `WriteOnlyCollections`, namely that there are some relationships that exist in the DB but shouldn't or often won't be expressed completely in the ORM, seems relevant, but as you say`WriteOnlyCollections` aren't populated; it's in the name. `Load.with_expression()` is capable of arbitrary loading, so maybe that is the tool to use, but it's so general that it has no specific help with loading relationships.

I'm envisioning a type of relationship loader that behaves like a `WriteOnlyCollection` for INSERTS, UPDATES, DELETES; like a `query_expression` for custom eager loading, but with a automatic `default_expr` that behaves like a standard relationship if there is no `with_expression()` given. I think an api along these lines would be a general solution to meet a general need, which, though admittedly not huge, is bigger than myself.

I don't know if the above sounds perceptive or oblivious. If you think I'm onto something; I'm happy to go back and forth some more about what this design space could look like, and to help code it. If you think what I've written is more confused than helpful, then I'll back off. Very grateful for SQLAlchemy regardless. :)

**Reply by CaselIT** [reply]
Since what you want to do with that graphql is load at most 100 object for each parent, ect, you can't really do that with eager load even if they were to support limit, since it's not a global query limit but a per-parent one.

Do do this in sql you would probably need a window function (likely rank or row_number) partitioning over the parent id. Sadly that can't be used in the where clause

**Reply by zzzeek** [reply]
> Yes, I understand that at present WriteOnlyCollections are never populated. What I'm saying is, perhaps they ought to be (though granted that would make the name a misnomer)?

ah...well, no, they were designed very explicitly, not even a year ago, to be exactly what they are!   having these arbitrary collections stuck onto the object just invites a lot of confusion and we'd prefer SQLAlchemy to be as minimal and unambiguous as possible.

I would recommend using a regular relationship and if you want to populate them explicitly exclusively, you can set lazy='noload' on them.


> "While loading several parents, I want to eagerly load their children, but subject to arbitrary WHERE/LIMIT/ORDER BY/etc. How do I do that?", then it doesn't seem there is a clear right tool. 

You're right, there's not, beyond the recipe I illustrated.   I have no idea how such a tool would be designed such that it's fully open-ended, unambiguous, and completely bug free.   I suppose *certain* eager loaders, and by "certain" I pretty much mean `selectinload()`, could be given additional optional parameters "limit" and "order_by".   But it starts looking like a frankenstein is being built - if we do that, then what if I want a CTE with my selectinload also?  it starts to look like we dug a hole into some part of the API when a doorway was already there (that is, use select() objects).

I think my suggestion to write your own queries and then populate collections directly as needed should at least be considered as the basis for a new tool that you could design and implement (and probably maintain, if you were to release it publicly...sorry we have a lot to maintain here already).

**Reply by CaselIT** [reply]
Not sure if maybe a public api to get the query that a loader would run (maybe just for selectinloader). Something like this
```
def get_selectin_for_instances(relationship, instances) -> Select: ...
```

This would create the select that would be run by the selectin (basically the `select * from rel_table where rel_table.parent_id in (ids of the instances)`). This would then be possible to run that query to as is or use as the start of a more complex query adding additional conditions / sort / limits.

**Reply by zzzeek** [reply]
I realized that LIMIT would be very difficult to do for selectinloader, and really would be very difficult for *all* the eager loaders except "immediateload", since they load all collections at once, requiring highly complicated and inefficient query forms for basically all of them.   so LIMIT would not be workable in any case.

**Reply by CaselIT** [reply]
yes, I wrote that here https://github.com/sqlalchemy/sqlalchemy/discussions/9898#discussioncomment-6097684

you would need a window function with partition to apply a limit to each parent.

This is something that could be added by user code to the select returned by an hypothetical `get_selectin_for_instances` though

**Reply by IohannesArnold** [reply]
Okay if `WriteOnlyCollection` is designed for a different purpose, then forget me bringing it up. And I see your point about punching holes in the wall where there's already a `select()` door. I just want to try one more line of thought.

Could there be a loader like `Load.customselect(attr, stmt: Select)`? This would be nearly the inverse of what Frederico is suggesting; someone would create a custom select statement that returned the right objects for `attr` but also did all the weird filtering/limiting/etc one desired, and then `Load.customselect` would add the `WHERE ... IN ...` condition just like a `selectinload`.

Also, I've learned a lot from this exchange and would like to expand the docs on this topic if you think it'd be useful. Is there any special procedure for submitting a pull request against the docs?

**Reply by zzzeek** [reply]
the thing is you can't generalize "selectinload" or any of the other eager loaders to achieve LIMIT per parent id without using some complex form of subqueries / window functions etc.    We would never be trying to generate SQL like that on this end, generating subqueries generically is always very complicated and error prone.

re: docs, in this case we should discuss what it is we want to document.

**Reply by CaselIT** [reply]
> Not sure if maybe a public api to get the query that a loader would run (maybe just for selectinloader). > Something like this
> `def get_selectin_for_instances(relationship, instances) -> Select: ...`

mike pointed out that there is `with_parent` that kinda does this

**Reply by IohannesArnold** [reply]
Okay, so I finally rolled up by sleeves and rewrote the my data-loading code as Mike suggested to use separate `select()` statements. It clearly is the right way to go; since I've made the change, the code has indeed been easier to go back and make further tweaks to. So my apologies for trying to bargain my way out of a code re-write. 😅 I appreciate your guidance.

I do want to offer to make what I think would be a helpful change to the docs for this situation. Right now there are two sections about filtering relationships: [Adding Criteria to loader options](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#adding-criteria-to-loader-options) and [Using contains_eager() to load a custom-filtered collection result](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#using-contains-eager-to-load-a-custom-filtered-collection-result), but these are under separate `<h2>`s. I think it would be better to have an additional `<h2>` entitled something like 'Filtering relationship collections', which began with a warning that this likely a code smell, then containing _Adding Criteria to loader options_ and _Using contains_eager() to load a custom-filtered collection result_, and finally pointing to the disjoint loading example for a custom implementation. If you'd like to see a PR for these changes to the docs, let me know.

Finally, not expecting the maintainers to look over my code, but for anyone that stumbles on this page in the future and wants to know what I did, here's what I implemented:
```python
def recursive_eager_load(stmt: Select, DBModelClass: Type[DeclarativeBase], gql_field):
    db_column_loads = []
    db_relationship_loads = []
    inspection = inspect(DBModelClass)

    for subfield in gql_field.selections:
        db_model_attr = getattr(DBModelClass, subfield.name, None)
        if db_model_attr is None or not hasattr(db_model_attr, "property"):
            continue
        elif isinstance(db_model_attr.property, ColumnProperty):
            db_column_loads.append(db_model_attr)
        elif isinstance(db_model_attr.property, Relationship):
            db_relationship_loads.append((db_model_attr, subfield))

    if len(db_column_loads) > 0:
        lo = load_only(*db_column_loads)
    else:
        lo = load_only(DBModelClass.id)
    rows = context.db_session.execute(stmt.options(lo)).all()

    for relationship_attr, gql_subfield in db_relationship_loads:
        load_relationship(inspection.class_, relationship_attr, gql_subfield, rows)

    return rows

def get_best_parent_id_col(relationship: MapperProperty, ParentDBClass: Type[DeclarativeBase]):
    for column in relationship.remote_side:
        if ParentDBClass.id in [fk.column for fk in column.foreign_keys]:
            return column
    return ParentDBClass.id

def load_relationship(
    ParentDBClass: Type[DeclarativeBase],
    relationship_attr: InstrumentedAttribute,
    gql_subfield: SelectedField,
    parent_rows: List[Row],
):
    ChildDBClass = relationship_attr.mapper.class_
    parent_ids = [row[0].id for row in parent_rows]

    parent_id_col = get_best_parent_id_col(relationship_attr.property, ParentDBClass)
    child_stmt = select(ChildDBClass, parent_id_col.label("parent_id"))

    if relationship_attr.property.secondary is not None:
        child_stmt = child_stmt.join(relationship_attr.property.secondary)
    elif parent_id_col == ParentDBClass.id and ParentDBClass == ChildDBClass:
        ChildDBClass = aliased(ChildDBClass)
        child_stmt = select(ChildDBClass, parent_id_col.label("parent_id"))
        child_stmt = child_stmt.join(ChildDBClass, relationship_attr)
    elif parent_id_col == ParentDBClass.id:
        child_stmt = child_stmt.join(relationship_attr)

    load_number = gql_subfield.arguments.get("first")

    child_stmt = child_stmt.where(parent_id_col.in_(parent_ids))

    if load_number:
        cte = child_stmt.add_columns(
            sql_func.row_number()
            .over(partition_by=parent_id_col, order_by=ChildDBClass.id)
            .label("rn")
        ).cte()
        ChildDBClass = aliased(ChildDBClass, cte)
        parent_id_col = cte.c.parent_id
        child_stmt = select(ChildDBClass, cte.c.parent_id).where(
            cte.c.rn <= int(load_number)
        )

    child_stmt = child_stmt.order_by(parent_id_col, ChildDBClass.id)

    child_rows = recursive_eager_load(info, child_stmt, ChildDBClass, gql_subfield)
    for k, g in groupby(child_rows, lambda row: row[1]):
        p = context.db_session.get(ParentDBClass, k)
        parent_ids.remove(k)
        gl = [row[0] for row in g]
        if relationship_attr.property.direction.name == "MANYTOONE":
            assert len(gl) == 1
            gl = gl[0]
        attributes.set_committed_value(p, relationship_attr.key, gl)

    # Handle the remaining parents that don't have any children
    for k in parent_ids:
        p = context.db_session.get(ParentDBClass, k)
        if relationship_attr.property.direction.name == "MANYTOONE":
            attributes.set_committed_value(p, relationship_attr.key, None)
        else:
            attributes.set_committed_value(p, relationship_attr.key, [])
```

**Reply by zzzeek** [reply]
I can see adding a top level "Tip:" box to https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#using-contains-eager-to-load-a-custom-filtered-collection-result that folks might not need this.  that's a good idea

**Reply by CaselIT** [reply]
I believe the tip is now live at that link

**Reply by IohannesArnold** [reply]
Okay, well that's all I have then. Thanks again for the help!
